File: src/main/java/com/ndgndg91/ordermatchingengine/order/service/Engine.java
package com.ndgndg91.ordermatchingengine.order.service;

import com.ndgndg91.ordermatchingengine.global.OrderServiceException;
import com.ndgndg91.ordermatchingengine.order.*;
import com.ndgndg91.ordermatchingengine.order.dto.request.AddOrderRequest;
import com.ndgndg91.ordermatchingengine.order.dto.request.CancelOrderRequest;
import com.ndgndg91.ordermatchingengine.order.dto.request.ModifyOrderRequest;
import com.ndgndg91.ordermatchingengine.order.event.OrderMatchTriggerEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.event.EventListener;
import org.springframework.http.HttpStatus;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class Engine {
    private final OrderBooks orderBooks;
    private final MatchedOrders matchedOrders;
    private final ApplicationEventPublisher publisher;

    public void addOrder(AddOrderRequest request) {
        final var orderID = UUID.randomUUID().toString();
        final var order = Order.builder()
                .orderId(orderID)
                .orderType(OrderType.valueOf(request.getOrderType()))
                .symbol(Symbol.valueOf(request.getSymbol()))
                .shares(request.getShares())
                .priceType(PriceType.valueOf(request.getPriceType()))
                .price(request.getPrice())
                .timestamp(LocalDateTime.now())
                .build();

        log.info("{}", order);
        orderBooks.addOrder(order);
        publisher.publishEvent(new OrderMatchTriggerEvent(order.getSymbol(), order.getPriceType(), order.getOrderType()));
    }

    public void modifyOrder(ModifyOrderRequest request) {
        final var order = Order.builder()
                .orderId(request.getOrderId())
                .orderType(OrderType.valueOf(request.getOrderType()))
                .symbol(Symbol.valueOf(request.getSymbol()))
                .shares(request.getShares())
                .priceType(PriceType.valueOf(request.getPriceType()))
                .price(request.getPrice())
                .timestamp(LocalDateTime.now())
                .build();
        orderBooks.modifyOrder(order);
        publisher.publishEvent(new OrderMatchTriggerEvent(order.getSymbol(), order.getPriceType(), order.getOrderType()));
    }

    public void cancelOrder(CancelOrderRequest request) {
        final var order = Order.builder()
                .orderId(request.getOrderId())
                .orderType(OrderType.valueOf(request.getOrderType()))
                .symbol(Symbol.valueOf(request.getSymbol()))
                .build();
        orderBooks.cancelOrder(order);
        publisher.publishEvent(new OrderMatchTriggerEvent(order.getSymbol(), order.getPriceType(), order.getOrderType()));
    }

    public Optional<OrderEntry> pollBids(Symbol symbol) {
        OrderBook orderBook = orderBooks.findOrderBooksBySymbol(symbol)
                .orElseThrow(() -> new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "not found order book by : " + symbol));

        return orderBook.bidsPoll();
    }

    public Optional<OrderEntry> pollAsks(Symbol symbol) {
        OrderBook orderBook = orderBooks.findOrderBooksBySymbol(symbol)
                .orElseThrow(() -> new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), "not found order book by : " + symbol));
        return orderBook.asksPoll();
    }

    public List<MatchResult> findAllMatchResults(Symbol symbol) {
        return matchedOrders.findAll(symbol);
    }

    @Async
    @EventListener
    public void match(OrderMatchTriggerEvent event) {
        log.info("{}", event);
        orderBooks.match(event.getSymbol(), event.getPriceType(), event.getOrderType())
                .ifPresent(matchedOrders::addMatchedOrder);
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/order/service/Engine.kt
package com.ndgndg91.ordermatchingenginekotlin.order.service

import com.ndgndg91.ordermatchingenginekotlin.order.*
import com.ndgndg91.ordermatchingenginekotlin.order.dto.request.AddOrderRequest
import com.ndgndg91.ordermatchingenginekotlin.order.dto.request.CancelOrderRequest
import com.ndgndg91.ordermatchingenginekotlin.order.dto.request.ModifyOrderRequest
import com.ndgndg91.ordermatchingenginekotlin.order.event.OrderMatchTriggerEvent
import com.ndgndg91.ordermatchingenginekotlin.order.repository.MatchOrderRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import java.util.*

@Service
class Engine(
    private val orderBooks: OrderBooks,
    private val matchOrderRepository: MatchOrderRepository,
    private val publisher: ApplicationEventPublisher
) {
    private val log: Logger = LoggerFactory.getLogger(Engine::class.java)
    fun addOrder(request: AddOrderRequest): Order {
        log.info("add order request : {}", request)
        val order = Order.Builder()
            .orderId(UUID.randomUUID().toString())
            .orderType(OrderType.valueOf(request.orderType!!))
            .priceType(PriceType.valueOf(request.priceType!!))
            .price(request.price!!)
            .symbol(Symbol.valueOf(request.symbol!!))
            .shares(request.shares!!)
            .build()
        log.info("add order : {}", order)
        orderBooks.addOrder(order)
        publisher.publishEvent(OrderMatchTriggerEvent(order.symbol, order.priceType, order.orderType))
        return order
    }

    fun modifyOrder(request: ModifyOrderRequest) {
        log.info("modify order request : {}", request)
        val order = Order.Builder()
            .orderId(request.orderId!!)
            .orderType(OrderType.valueOf(request.orderType!!))
            .priceType(PriceType.valueOf(request.priceType!!))
            .price(request.price!!)
            .symbol(Symbol.valueOf(request.symbol!!))
            .shares(request.shares!!)
            .build()
        log.info("modify order : {}", order)
        orderBooks.modifyOrder(order)
        publisher.publishEvent(OrderMatchTriggerEvent(order.symbol, order.priceType, order.orderType))
    }

    fun cancelOrder(request: CancelOrderRequest) {
        log.info("cancel order request : {}", request)
        val order = Order.Builder()
            .orderId(request.orderId!!)
            .orderType(OrderType.valueOf(request.orderType!!))
            .symbol(Symbol.valueOf(request.symbol!!))
            .build()
        log.info("cancel order : {}", order)
        publisher.publishEvent(OrderMatchTriggerEvent(order.symbol, order.priceType, order.orderType))
        orderBooks.cancelOrder(order)
    }

    fun find(symbol: String, orderType: String, orderId: String): OrderEntry? {
        return orderBooks.find(Symbol.valueOf(symbol), OrderType.valueOf(orderType), orderId)
    }

    fun test(symbol: String): MutableList<String>? {
        return matchOrderRepository.findAllBySymbol(symbol)
    }

    @Async
    @EventListener
    fun match(event: OrderMatchTriggerEvent) {
        log.info("{}", event)
        orderBooks.match(event.symbol, event.priceType, event.orderType)
            .also { log.info("$it"); }
            .let { matchOrderRepository.save(event.symbol, it) }
    }
}